api/graphite.go

```diff
@@ -888,28 +889,25 @@ func (s *Server) graphiteTagFindSeries(ctx *middleware.Context, request models.G
 	response.Write(ctx, response.NewJson(200, seriesNames, ""))
 }
 
-func (s *Server) clusterFindByTag(ctx context.Context, orgId uint32, expressions []string, from int64) ([]Series, error) {
+func (s *Server) clusterFindByTag(ctx context.Context, orgId uint32, expressions []string, from int64, maxSeries int) ([]Series, error) {
 	data := models.IndexFindByTag{OrgId: orgId, Expr: expressions, From: from}
-	resps, err := s.peerQuerySpeculative(ctx, data, "clusterFindByTag", "/index/find_by_tag")
-	if err != nil {
-		return nil, err
-	}
-
-	select {
-	case <-ctx.Done():
-		//request canceled
-		return nil, nil
-	default:
-	}
+	responseChan, errorChan := s.peerQuerySpeculativeChan(ctx, data, "clusterFindByTag", "/index/find_by_tag")
 
 	var allSeries []Series
 
-	for _, r := range resps {
+	for r := range responseChan {
 		resp := models.IndexFindByTagResp{}
-		_, err = resp.UnmarshalMsg(r.buf)
+		_, err := resp.UnmarshalMsg(r.buf)
 		if err != nil {
 			return nil, err
 		}
+
+		// 0 disables the check, so only check if maxSeriesPerReq > 0
+		if maxSeriesPerReq > 0 && len(resp.Metrics)+len(allSeries) > maxSeries {
+			return nil,
```

Comment:
Shouldn't you cancel `ctx` so outstanding peer queries are aborted?

Reply:
Technically, the query will be `Done`d when the error propagates up and a response is returned. But canceling it here would do it a few milliseconds earlier.

Reply:
I'm also not entirely sure how to cancel this context here. I suppose we would need to make a new cancelable context here.

Reply:
that makes sense I think. Since we only get the ctx passed in, we don't know where what the cancelfunc was.
However, if you do something like `newCtx, cancel := context.WithCancel(ctx)` and pass newCtx into peerQuerySpeculativeChan, then you can cancel it.